runtime/sema/checker.go

```diff
@@ -381,6 +389,40 @@ func (checker *Checker) IsChecked() bool {
 	return checker.isChecked
 }
 
+func (checker *Checker) CurrentPurityScope() PurityCheckScope {
+	return checker.purityCheckScopes[len(checker.purityCheckScopes)-1]
+}
+
+func (checker *Checker) PushNewPurityScope(enforce bool) {
+	checker.purityCheckScopes = append(checker.purityCheckScopes, PurityCheckScope{EnforcePurity: enforce, CurrentPurity: true})
+}
+
+func (checker *Checker) PopPurityScope() PurityCheckScope {
+	scope := checker.CurrentPurityScope()
+	checker.purityCheckScopes = checker.purityCheckScopes[:len(checker.purityCheckScopes)-1]
+	return scope
+}
+
+func (checker *Checker) ObserveImpureOperation(operation ast.Element) {
+	scope := checker.CurrentPurityScope()
+	// purity is monotonic, if we already know this scope is impure, there's no need to continue
+	if !scope.CurrentPurity {
```

Comment:
Ideally this should keep the performance of the analysis within the realm of reason since we won't be recording every single impure event, only the first one in each function body.